pkg/fanal/analyzer/language/dart/pub/pubspec.go

```diff
@@ -2,35 +2,177 @@ package pub
+			})
+
+			for i, lib := range app.Libraries {
+				var dependsOn []string
+				for _, depName := range allDependsOn[lib.ID] {
+					if depID, ok := libs[depName]; ok {
+						dependsOn = append(dependsOn, depID)
+					}
+				}
+				app.Libraries[i].DependsOn = dependsOn
+			}
+		}
+
+		sort.Sort(app.Libraries)
+		apps = append(apps, *app)
+		return nil
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("walk error: %w", err)
+	}
+
+	return &analyzer.AnalysisResult{
+		Applications: apps,
+	}, nil
+}
 
-func (a pubSpecLockAnalyzer) Analyze(_ context.Context, input analyzer.AnalysisInput) (*analyzer.AnalysisResult, error) {
-	p := pub.NewParser()
-	res, err := language.Analyze(types.Pub, input.FilePath, input.Content, p)
+func findDependsOn() (map[string][]string, error) {
+	dir := cacheDir()
+	if !fsutils.DirExists(dir) {
+		log.Logger.Debugf("Cache dir (%s) not found. Need 'dart pub get' to fill dependency relationships", dir)
+		return nil, nil
+	}
+
+	deps := make(map[string][]string)
+	if err := filepath.WalkDir(dir, func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		} else if !d.Type().IsRegular() {
+			return nil
+		}
+		// parse only `pubspec.yaml` files
+		if filepath.Base(p) != pubSpecYamlFileName {
+			return nil
+		}
+
+		id, dependsOn, err := parsePubSpecYaml(p)
+		if err != nil {
+			log.Logger.Debugf("unable to parse %q: %s", p, err)
+			return nil
+		}
+		if id != "" {
+			deps[id] = dependsOn
+		}
+		return nil
+
+	}); err != nil {
+		return nil, xerrors.Errorf("walk error: %w", err)
+	}
+	return deps, nil
+}
+
+// https://dart.dev/tools/pub/glossary#system-cache
+func cacheDir() string {
+	if dir := os.Getenv("PUB_CACHE"); dir != "" {
+		return dir
+	}
+
+	// `%LOCALAPPDATA%\Pub\Cache` for Windows
+	if runtime.GOOS == "windows" {
+		return filepath.Join(os.Getenv("LOCALAPPDATA"), "Pub", "Cache")
+	}
+
+	// `~/.pub-cache` for Linux or Mac
+	return filepath.Join(os.Getenv("HOME"), ".pub_cache")
+}
+
+type pubSpecYaml struct {
+	Name         string                 `yaml:"name"`
+	Version      string                 `yaml:"version"`
+	Dependencies map[string]interface{} `yaml:"dependencies,omitempty"`
+}
+
+func parsePubSpecYaml(filePath string) (string, []string, error) {
+	f, err := os.Open(filePath)
 	if err != nil {
-		return nil, xerrors.Errorf("%s parse error: %w", input.FilePath, err)
+		return "", nil, xerrors.Errorf("unable to open %q to get list of direct deps: %w", filePath, err)
+	}
+	defer func() { _ = f.Close() }()
+
+	var spec pubSpecYaml
+	if err = yaml.NewDecoder(f).Decode(&spec); err != nil {
+		return "", nil, xerrors.Errorf("unable to decode %q: %w", filePath, err)
+	}
+	if len(spec.Dependencies) > 0 {
+		// pubspec.yaml uses version ranges
+		// save only dependencies names
+		dependsOn := lo.MapToSlice(spec.Dependencies, func(key string, _ interface{}) string {
+			return key
+		})
+		return utils.PackageID(spec.Name, spec.Version), dependsOn, nil
```

Comment:
It is better to align the happy path to the left.

```suggestion
	if len(spec.Dependencies) == 0 {
		return "", nil, nil
```

Reply:
Fixed in https://github.com/aquasecurity/trivy/pull/5374/commits/ffe7308f7258d614b98ec9752b01889944feb628